fact/src/cgroup.rs

```diff
@@ -0,0 +1,174 @@
+use std::{
+    collections::HashMap,
+    os::unix::fs::DirEntryExt,
+    path::PathBuf,
+    sync::Arc,
+    time::{Duration, SystemTime},
+};
+
+use log::warn;
+use tokio::{
+    sync::{watch::Receiver, Mutex},
+    task::JoinHandle,
+    time,
+};
+
+use crate::host_info::get_cgroup_paths;
+
+#[derive(Debug)]
+struct ContainerIdEntry {
+    container_id: Option<Arc<String>>,
+    pub last_seen: SystemTime,
+}
+
+type ContainerIdMap = HashMap<u64, ContainerIdEntry>;
+
+#[derive(Debug, Clone, Default)]
+pub struct ContainerIdCache(Arc<Mutex<ContainerIdMap>>);
+
+impl ContainerIdCache {
+    pub fn new() -> Self {
+        let mut map = HashMap::new();
+        ContainerIdCache::update_unlocked(&mut map);
+        ContainerIdCache(Arc::new(Mutex::new(map)))
+    }
+
+    fn update_unlocked(map: &mut ContainerIdMap) {
```

Comment:
Maybe rename to `update_no_lock` or `update_without_lock`. The name implies that this updates something that is unlocked, rather than doing an update without locking.